stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,277 @@
+        # which is "output".Iinstead, output_names is ["concatenate_4"].
+        # output_names = model_csbdeep.output_names
+        output_names = ["output"]
+
+        output_n_channels = (1 + model.config.n_rays,)
+        output_scale = [1]*(ndim_tensor-1) + [0]
+
+    # TODO need config format that is compatible with deepimagej; discuss with Esti
+    # TODO do we need parameters for down/upsampling here?
+    package_data = metadata("stardist")
+    config = dict(
+        stardist=dict(
+            stardist_version=package_data["Version"],
+            thresholds=dict(nms=model.thresholds.nms, prob=model.thresholds.prob)
+        )
+    )
+
+    n_inputs = len(input_names)
+    assert n_inputs == 1
+    input_axes = "b" + net_axes_in.lower()
+    input_config = dict(
+        input_name=input_names,
+        input_step=[[0]+div_by] * n_inputs,
+        input_min_shape=[[1] + div_by] * n_inputs,
+        input_axes=[input_axes] * n_inputs,
+        input_data_range=[["-inf", "inf"]] * n_inputs,
+        preprocessing=[dict(scale_range=dict(
+            mode="per_sample",
+            # TODO mighe make it an option to normalize across channels ...
+            axes=net_axes_in.lower().replace("c", ""),
+            min_percentile=min_percentile,
+            max_percentile=max_percentile,
+        ))] * n_inputs
+    )
+
+    n_outputs = len(output_names)
+    assert len(output_n_channels) == n_outputs
+    output_axes = "b" + net_axes_out.lower()
+    output_config = dict(
+        output_name=output_names,
+        output_data_range=[["-inf", "inf"]] * n_outputs,
+        output_axes=[output_axes] * n_outputs,
+        output_reference=[input_names[0]] * n_outputs,
+        output_scale=[output_scale] * n_outputs,
+        output_offset=[[1] * (ndim_tensor-1) + [n_channel] for n_channel in output_n_channels]
+    )
+
+    in_path = outdir / "test_input.npy"
+    np.save(in_path, _expand_dims(test_input, input_axes))
+
+    test_input = normalize(test_input, pmin=min_percentile, pmax=max_percentile)
+    if mode == "tensorflow_saved_model_bundle":
+        test_outputs = _predict_tf(weight_uri, _expand_dims(test_input, input_axes))
+    else:
+        test_outputs = model.predict(test_input)
+
+    out_paths = []
+    for i, out in enumerate(test_outputs):
+        p = outdir / f"test_output{i}.npy"
+        np.save(p, _expand_dims(out, output_axes))
+        out_paths.append(p)
+
+    data = dict(weight_uri=weight_uri, test_inputs=[in_path], test_outputs=out_paths, config=config)
+    data.update(input_config)
+    data.update(output_config)
+    return data
+
+
+def export_bioimageio(
+    model,
+    outpath,
+    test_input,
+    name=None,
+    mode="tensorflow_saved_model_bundle",
+    prefer_weights="best",
+    min_percentile=1.0,
+    max_percentile=99.8,
+    overwrite_spec_kwargs={}
```

Comment:
This seems to be an anti-pattern in Python, see here: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Perhaps it's better to do `**kwargs` for extra kwargs?

Reply:
I agree in general. The function here is based on an earlier PR from @maweigert though, and there `overwrite_spec_kwargs` are given as dict, see https://github.com/stardist/stardist/blob/modelzoo/stardist/export_bioimageio.py#L207.
That's why I kept this pattern here assuming there is some reason for it. Happy to change it to proper kwargs.

Reply:
I agree, that the inital value should be changed to `None` to avoid potential mutability side effects, but would leave the name as is to indicate that these are not keyword parameters of the function but of the specs.

Reply:
@maweigert Wei is suggesting to use `**overwrite_spec_kwargs` so that the function takes arbitrary additional keyword arguments.

Reply:
Advantage of leaving it as an explicit dict argument: it allows overwriting spec attributes with the same name as the current function params

Reply:
> Advantage of leaving it as an explicit dict argument: it allows overwriting spec attributes with the same name as current function parameters

That's true. I am happy with keeping this as a dict and changing the default to `None`.